src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2370 @@
+    mi->state_length = CIRCPAD_STATE_LENGTH_INFINITE;
+    return;
+  }
+
+  length = circpad_distribution_sample(state->length_dist);
+  length = MAX(0, length);
+  length += state->start_length;
+  length = MIN(length, state->max_length);
+
+  mi->state_length = clamp_double_to_int64(length);
+}
+
+/**
+ * Sample an expected time-until-next-packet delay from the histogram.
+ *
+ * The bin is chosen with probability proportional to the number
+ * of tokens in each bin, and then a time value is chosen uniformly from
+ * that bin's [start,end) time range.
+ */
+static circpad_delay_t
+circpad_machine_sample_delay(circpad_machineinfo_t *mi)
+{
+  const circpad_state_t *state = circpad_machine_current_state(mi);
+  const circpad_hist_bin_t *histogram = NULL;
+  int i = 0;
+  uint32_t curr_weight = 0;
+  uint32_t histogram_total = 0;
+  uint32_t bin_choice;
+  circpad_delay_t bin_start, bin_end;
+  circpad_delay_t start_usec;
+
+  tor_assert(state);
+
+  if (state->use_rtt_estimate)
+    start_usec = mi->rtt_estimate_us+state->start_usec;
+  else
+    start_usec = state->start_usec;
+
+  if (state->token_removal != CIRCPAD_TOKEN_REMOVAL_NONE) {
+    /* We have a mutable histogram */
+    tor_assert(mi->histogram && mi->histogram_len == state->histogram_len);
+
+    histogram = mi->histogram;
+    for (int b = 0; b < state->histogram_len; b++)
+      histogram_total += histogram[b];
+  } else if (state->iat_dist.type == CIRCPAD_DIST_NONE) {
+    /* We have a histogram, but it's immutable */
+    histogram = state->histogram;
+    histogram_total = state->histogram_total;
+  } else {
+    /* Sample from a fixed IAT distribution and return */
+    double val = circpad_distribution_sample(state->iat_dist);
+    /* These comparisons are safe, because the output is in the range
+     * [0, 2**46], and double has a precision of 53 bits. */
+    val = MAX(0, val);
+    val = MIN(val, state->range_sec*TOR_USEC_PER_SEC);
+
+    /* This addition is exact: val is at most 2**16 * USEC_PER_SEC ~= 2**46,
+     * start_usec is at most 2**32, and doubles have a precision of 53 bits. */
+    val += start_usec;
+    /* Clamp the distribution at infinite delay val */
+    return (circpad_delay_t)MIN(tor_llround(val), CIRCPAD_DELAY_INFINITE);
+  }
+
+  bin_choice = crypto_rand_int(histogram_total);
+
+  /* Skip all the initial zero bins */
+  while (!histogram[i]) {
+    i++;
+  }
+  curr_weight = histogram[i];
+
+  // TODO: This is not constant-time. Pretty sure we don't
+  // really need it to be, though.
+  while (curr_weight < bin_choice) {
+    i++;
+    tor_assert(i < state->histogram_len);
+    curr_weight += histogram[i];
+  }
+
+  tor_assert(i < state->histogram_len);
+  tor_assert(histogram[i] > 0);
+
+  // Store this index to remove the token upon callback.
+  if (state->token_removal != CIRCPAD_TOKEN_REMOVAL_NONE) {
+    mi->chosen_bin = i;
+  }
+
+  if (i == state->histogram_len-1) {
+    if (state->token_removal != CIRCPAD_TOKEN_REMOVAL_NONE &&
```

Comment:
> I'm a bit confuse on how the token removal logic works. I don't think it corresponds to what `3.1.2` in `prop254` says. In particular, why is there an explicit rule here about removing tokens from the infinity bin? Why is this not handled in the token removal functions?

The infinity bin has to have its tokens removed as soon as "infinite" delay is chosen. This is because there is no callback for it. For all other bins, tokens are removed upon the padding callback, or if a non-padding cell is sent *before* the callback. We could have defined the infinity token removal to happen only if no non-padding cell ever is sent.. But those semantics are both more complicated, and not as useful for having a fixed number of "infinite" padding delays in a state. Does this make sense?

> Furthermore, we also remove tokens from bins in `circpad_send_padding_cell_for_callback()` regardless of what our token removal strategy is, even tho the spec says `If token removal is enabled, when a padding packet is sent, a token is removed from the bin corresponding to the target delay.`.

This is correct, and the implementation does in fact match the spec. So long as the token removal strategy is not NONE, then we will have a mutable histogram in the machine info. For all token removal strategies other than NONE, the padding case is the same -- remove from exactly the bin corresponding to the chosen delay. This bin should not be empty by definition -- otherwise padding could not have been scheduled for this bin's delay value.. Hence the asserts before token removal upon actually sending the padding cell.

> Also, the spec seems to imply that special token removal rules only apply when the exact bin has no tokens left. I don't see where this logic is in the code. I just see us applying the token removal rule, when we send a non-padding cell.

Every token removal rule other than NONE will try to remove from exactly the bin corresponding to the delay value that passed. It is only if this bin is empty do they apply their strategy. I could see an argument for factoring out this case and only applying the strategy-specific code if the target bin is in fact empty, but this mans either duplicating the target bin calculations, or passing the target bin weirdly between a bunch of the helper functions...